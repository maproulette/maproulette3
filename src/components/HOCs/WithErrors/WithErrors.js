import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import {
  addError,
  addErrorWithDetails,
  clearErrors,
  removeError,
} from "../../../services/Error/Error";

export const mapStateToProps = (state) => {
  return { errors: state.currentErrors };
};

export const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      addError,
      addErrorWithDetails,
      removeError,
      clearErrors,
    },
    dispatch,
  );

const WithErrors = (WrappedComponent) =>
  connect(mapStateToProps, mapDispatchToProps)(WrappedComponent);

export default WithErrors;
