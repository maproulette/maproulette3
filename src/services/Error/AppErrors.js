import messages from './Messages'

/**
 * App errors. Example use: `addError(AppErrors.user.unauthorized)`
 *
 * > Note, messages are kept separate so they'll be easily picked up when
 * > generating translation files.
 */
export default {
  user: {
    missingHomeLocation: messages.userMissingHomeLocation,
    unauthenticated: messages.userUnauthenticated,
    unauthorized: messages.userUnauthorized,
    updateFailure: messages.userUpdateFailure,
    fetchFailure: messages.userFetchFailure,
    notFound: messages.userNotFound,
  },

  leaderboard: {
    fetchFailure: messages.leaderboardFetchFailure,
  },

  task: {
    none: messages.taskNone,
    saveFailure: messages.taskSaveFailure,
    updateFailure: messages.taskUpdateFailure,
    deleteFailure: messages.taskDeleteFailure,
    fetchFailure: messages.taskFetchFailure,
    doesNotExist: messages.taskDoesNotExist,
    locked: messages.taskLocked,
    lockRefreshFailure: messages.taskLockRefreshFailure,
    bundleFailure: messages.taskBundleFailure,
  },

  osm: {
    requestTooLarge: messages.osmRequestTooLarge,
    bandwidthExceeded: messages.osmBandwidthExceeded,
    elementMissing: messages.osmElementMissing,
    fetchFailure: messages.osmFetchFailure,
  },

  mapillary: {
    fetchFailure: messages.mapillaryFetchFailure,
  },

  nominatim: {
    fetchFailure: messages.nominatimFetchFailure,
  },

  clusteredTask: {
    fetchFailure: messages.clusteredTaskFetchFailure,
  },

  boundedTask: {
    fetchFailure: messages.boundedTaskFetchFailure,
  },

  reviewTask: {
    fetchFailure: messages.reviewTaskFetchFailure,
    alreadyClaimed: messages.reviewTaskAlreadyClaimed,
  },

  challenge: {
    fetchFailure: messages.challengeFetchFailure,
    searchFailure: messages.challengeSearchFailure,
    deleteFailure: messages.challengeDeleteFailure,
    saveFailure: messages.challengeSaveFailure,
    rebuildFailure: messages.challengeRebuildFailure,
    doesNotExist: messages.challengeDoesNotExist,
  },

  virtualChallenge: {
    fetchFailure: messages.virtualChallengeFetchFailure,
    createFailure: messages.virtualChallengeCreateFailure,
    expired: messages.virtualChallengeExpired,
  },

  project: {
    saveFailure: messages.projectSaveFailure,
    fetchFailure: messages.projectFetchFailure,
    searchFailure: messages.projectSearchFailure,
    deleteFailure: messages.projectDeleteFailure,
    notManager: messages.projectNotManager,
  },

  map: {
    renderFailure: messages.mapRenderFailure,
    placeNotFound: messages.placeNotFound,
  },

  widgetWorkspace: {
    renderFailure: messages.widgetWorkspaceRenderFailure,
    importFailure: messages.widgetWorkspaceImportFailure,
  },

  josm: {
    noResponse: messages.josmNoResponse,
    missingOSMIds: messages.josmMissingOSMIds,
  },
}
