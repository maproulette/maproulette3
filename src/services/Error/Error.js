import PropTypes from 'prop-types'
import _isString from 'lodash/isString'
import _clone from 'lodash/clone'
import _isEmpty from 'lodash/isEmpty'
import _remove from 'lodash/remove'

export const errorShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  defaultMessage: PropTypes.string,
})

// redux actions
export const ADD_ERROR = 'ADD_ERROR'
export const REMOVE_ERROR = 'REMOVE_ERROR'
export const CLEAR_ERRORS = 'CLEAR_ERRORS'

/**
 * Build an error suitable for passing to addError. Errors should have id and
 * optional defaultMessage strings, which correspond to the id and
 * defaultMessage fields utilized by react-intl for displaying
 * internationalized messages. An optional detailMessage can be provided which,
 * if defined, will be appended on to the defaultMessage.
 */
export const buildError = function(id, defaultMessage, detailMessage) {
  if (!_isString(id) || id.length === 0) {
    throw new Error('Error id must be a valid string.')
  }

  const message = !_isEmpty(detailMessage) ?
    `${defaultMessage}: ${detailMessage}` :
    defaultMessage

  return {id, defaultMessage: message}
}

/**
 * Convenience method that delegates to buildError and attempts to extract an
 * error message from the given serverError and use it as the detailed message.
 * Note that this method returns a Promise that resolves with the built error
 * object.
 *
 * @returns a Promise that resolves with the built error.
 */
export const buildServerError = function(id, defaultMessage, serverError) {
  return new Promise(resolve => {
    if (serverError && serverError.response) {
      serverError.response.json().then(json => {
        resolve(buildError(id, defaultMessage,
                           _isString(json.message) ? json.message : undefined))
      }).catch(error => {
        resolve(buildError(id, defaultMessage))
      })
    }
    else {
      resolve(buildError(id, defaultMessage))
    }
  })
}


// redux action creators
export const addError = function(error) {
  return {
    type: ADD_ERROR,
    error,
  }
}

export const removeError = function(error) {
  return {
    type: REMOVE_ERROR,
    error,
  }
}

export const clearErrors = function() {
  return {
    type: CLEAR_ERRORS,
  }
}

// redux reducer
export const currentErrors = function(state=[], action) {
  let copy = null

  switch(action.type) {
    case ADD_ERROR:
      copy = _clone(state)
      copy.push(action.error)
      return copy
    case REMOVE_ERROR:
      copy = _clone(state)
      _remove(copy, {id: action.error.id})
      return copy
    case CLEAR_ERRORS:
      return []
    default:
      return state
  }
}
