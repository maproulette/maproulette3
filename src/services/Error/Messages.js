import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with AppErrors.
 */
export default defineMessages({
  userMissingHomeLocation: {
    id: "Errors.user.missingHomeLocation",
    defaultMessage: "No home location found. Please set your home location " +
                    "in your openstreetmap.org settings and then refresh this " +
                    "page to try again.",
  },
  userUnauthorized: {
    id: "Errors.user.unauthorized",
    defaultMessage: "Please sign in to continue.",
  },
  userUpdateFailure: {
    id: "Errors.user.updateFailure",
    defaultMessage: "Unable to update your user on server."
  },
  userFetchFailure: {
    id: "Errors.user.fetchFailure",
    defaultMessage: "Unable to fetch user data from server."
  },
  userNotFound: {
    id: "Errors.user.notFound",
    defaultMessage: "No user found with that username."
  },

  leaderboardFetchFailure: {
    id: "Errors.leaderboard.fetchFailure",
    defaultMessage: "Unable to fetch leaderboard."
  },

  taskNone: {
    id: 'Errors.task.none',
    defaultMessage: "No tasks remain in this challenge.",
  },
  taskSaveFailure: {
    id: 'Errors.task.saveFailure',
    defaultMessage: "Unable to save your changes{details}",
  },
  taskUpdateFailure: {
    id: 'Errors.task.updateFailure',
    defaultMessage: "Unable to save your changes.",
  },
  taskDeleteFailure: {
    id: 'Errors.task.deleteFailure',
    defaultMessage: "Unable to delete task.",
  },
  taskFetchFailure: {
    id: 'Errors.task.fetchFailure',
    defaultMessage: "Unable to fetch a task to work on.",
  },
  taskDoesNotExist: {
    id: 'Errors.task.doesNotExist',
    defaultMessage: "That task does not exist.",
  },

  clusteredTaskFetchFailure: {
    id: 'Errors.clusteredTask.fetchFailure',
    defaultMessage: "Unable to fetch task clusters",
  },

  boundedTaskFetchFailure: {
    id: 'Errors.boundedTask.fetchFailure',
    defaultMessage: "Unable to fetch map-bounded tasks",
  },

  challengeFetchFailure: {
    id: 'Errors.challenge.fetchFailure',
    defaultMessage: "Unable to retrieve latest challenge data from server."
  },
  challengeSearchFailure: {
    id: 'Errors.challenge.searchFailure',
    defaultMessage: "Unable to search challenges on server.",
  },
  challengeDeleteFailure: {
    id: 'Errors.challenge.deleteFailure',
    defaultMessage: "Unable to delete challenge.",
  },
  challengeSaveFailure: {
    id: 'Errors.challenge.saveFailure',
    defaultMessage: "Unable to save your changes{details}",
  },
  challengeRebuildFailure: {
    id: 'Errors.challenge.rebuildFailure',
    defaultMessage: "Unable to rebuild challenge tasks",
  },
  challengeDoesNotExist: {
    id: 'Errors.challenge.doesNotExist',
    defaultMessage: "That challenge does not exist.",
  },

  virtualChallengeFetchFailure: {
    id: 'Errors.virtualChallenge.fetchFailure',
    defaultMessage: "Unable to retrieve latest virtual challenge data from server."
  },
  virtualChallengeCreateFailure: {
    id: 'Errors.virtualChallenge.createFailure',
    defaultMessage: "Unable to create a virtual challenge{details}",
  },
  virtualChallengeExpired: {
    id: 'Errors.virtualChallenge.expired',
    defaultMessage: "Virtual challenge has expired.",
  },

  projectSaveFailure: {
    id: 'Errors.project.saveFailure',
    defaultMessage: "Unable to save your changes{details}",
  },
  projectFetchFailure: {
    id: 'Errors.project.fetchFailure',
    defaultMessage: "Unable to retrieve latest project data from server.",
  },
  projectSearchFailure: {
    id: 'Errors.project.searchFailure',
    defaultMessage: "Unable to search projects.",
  },
  projectDeleteFailure: {
    id: 'Errors.project.deleteFailure',
    defaultMessage: "Unable to delete project.",
  },
  projectNotManager: {
    id: 'Errors.project.notManager',
    defaultMessage: "You must be a manager of that project to proceed.",
  },

  mapRenderFailure: {
    id: 'Errors.map.renderFailure',
    defaultMessage: "Unable to render the map. This is usally caused by a problematic custom basemap.",
  },

  josmNoResponse: {
    id: 'Errors.josm.noResponse',
    defaultMessage: "OSM remote control did not respond. Do you have JOSM " +
                    "running with Remote Control enabled?"
  },
})
