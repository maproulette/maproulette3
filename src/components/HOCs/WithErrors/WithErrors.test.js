import { vi } from "vitest";
import { addError, clearErrors } from "../../../services/Error/Error";
import { mapDispatchToProps, mapStateToProps } from "./WithErrors";

vi.mock("../../../services/Error/Error");

let basicState = null;
let anError = null;

beforeEach(() => {
  anError = {
    id: "Errors.foo.bar.baz",
    defaultMessage: "A foo bar baz error.",
  };

  basicState = {
    currentErrors: [anError],
  };
});

test("mapStateToProps maps currentErrors to errors", () => {
  const mappedProps = mapStateToProps(basicState);
  expect(mappedProps.errors).toEqual(basicState.currentErrors);

  expect(mappedProps).toMatchSnapshot();
});

test("mapDispatchToProps maps function addError", () => {
  const dispatch = vi.fn(() => Promise.resolve());
  const mappedProps = mapDispatchToProps(dispatch, {});

  mappedProps.addError(anError);
  expect(dispatch).toBeCalled();
  expect(addError).toBeCalledWith(anError);
});

test("mapDispatchToProps maps function clearErrors", () => {
  const dispatch = vi.fn(() => Promise.resolve());
  const mappedProps = mapDispatchToProps(dispatch, {});

  mappedProps.clearErrors();
  expect(dispatch).toBeCalled();
  expect(clearErrors).toBeCalled();
});
