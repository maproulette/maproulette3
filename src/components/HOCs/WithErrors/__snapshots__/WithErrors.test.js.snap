// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mapStateToProps maps currentErrors to errors 1`] = `
{
  "errors": [
    {
      "defaultMessage": "A foo bar baz error.",
      "id": "Errors.foo.bar.baz",
    },
  ],
}
`;
